File: 19. Fruit Into Baskets.cpp
// Topic: Fruit Into Baskets (longest subarray with at most two distinct types)
#include <bits/stdc++.h>
using namespace std;

int totalFruit(vector<int>& fruits)
{
    unordered_map<int, int> cnt;
    int left = 0, res = 0;
    for (int right = 0; right < fruits.size(); ++right)
    {
        cnt[fruits[right]]++;
        while (cnt.size() > 2)
        {
            if (--cnt[fruits[left]] == 0)
                cnt.erase(fruits[left]);
            left++;
        }
        res = max(res, right - left + 1);
    }
    return res;
}

int main()
{
    vector<int> fruits = {1, 2, 1};
    cout << "Max fruits: " << totalFruit(fruits) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) or O(m) where m distinct types (here ≤ 2)

File: 19. Fruit Into Baskets.py
# Topic: Fruit Into Baskets (longest subarray with at most two distinct types)

from collections import defaultdict

def total_fruit(fruits):
    cnt = defaultdict(int)
    left = 0
    res = 0
    for right, f in enumerate(fruits):
        cnt[f] += 1
        while len(cnt) > 2:
            cnt[fruits[left]] -= 1
            if cnt[fruits[left]] == 0:
                del cnt[fruits[left]]
            left += 1
        res = max(res, right-left+1)
    return res

if __name__ == '__main__':
    fruits = [1,2,1]
    print('Max fruits:', total_fruit(fruits))

# Time Complexity: O(n)
# Space Complexity: O(1) (bounded by 2 types)
